File: stack/stack.go
/*
 ** Stack
 ** New - new Stack(value)
 ** Push - .push(value)
 ** Pop - .pop()
 ** Peek - .peek()
 */

// =============================================
package main

import "fmt"

type stack []int

func newStack(value int) stack {
	return stack{value}
}

func (s *stack) push(value int) {
	*s = append(*s, value)
}

func (s *stack) pop() int {
	top := (*s)[len(*s)-1]
	*s = (*s)[:len(*s)-1]
	return top
}

func (s stack) peek() int {
	return s[len(s)-1]
}

func main() {
	s := newStack(1)
	fmt.Println(s)
	s.push(10)
	fmt.Println(s)
	fmt.Println(s.pop())
	fmt.Println(s.peek())
	fmt.Println(s)
}

File: stack/stack.js
/*
 ** Stack
 ** New - new Stack(value)
 ** Push - .push(value)
 ** Pop - .pop()
 ** Peek - .peek()
 */

// =============================================

class Stack {
  constructor(value) {
    this.stack = [value];
  }

  peek() {
    if (this.stack.length > 0) {
      return this.stack[this.stack.length - 1];
    } else {
      console.log('Stack is empty!');
      return null;
    }
  }

  push(value) {
    this.stack.push(value);
  }

  pop() {
    if (this.stack.length > 0) {
      return this.stack.pop();
    } else {
      console.log('Stack is empty!');
      return null;
    }
  }
}

const s = new Stack(1);
console.log(s.stack);
s.push(10);
console.log(s.stack);
console.log(s.pop());
console.log(s.peek());
console.log(s.stack);
